AtReconstruction/AtFitter/OpenKF/kalman_filter/TrackFitterUKF.h

```diff
@@ -0,0 +1,375 @@
+/// https://opensource.org/licenses/GPL-3.0
+///
+/// @author Mohanad Youssef <mohanad.magdy.hammad@gmail.com>
+/// @file unscented_kalman_filter.h
+///
+
+#ifndef UNSCENTED_KALMAN_FILTER_LIB_H
+#define UNSCENTED_KALMAN_FILTER_LIB_H
+
+#include "kalman_filter.h"
+#include "kf_util.h"
+#include "unscented_kalman_filter.h"
+
+namespace kf {
+
+/// @brief Class for fitting tracks using the Unscented Kalman Filter (UKF) algorithm.
+/// @tparam DIM_X Dimension of the state vector.
+/// @tparam DIM_Z Dimension of the measurement vector.
+/// @tparam DIM_V Dimension of the process noise vector.
+/// @tparam DIM_N Dimension of the measurement noise vector.
+template <int32_t DIM_X = 6, int32_t DIM_Z = 3, int32_t DIM_V = 2, int32_t DIM_N = 3>
+class TrackFitterUKF : public KalmanFilter<DIM_X, DIM_Z> {
+public:
+   // Augmented state vector is just the process noise and state vector. The measurement noise is not included as that
+   // is independent of the propagation and measurement model and just adds linearly.
+   static constexpr int32_t DIM_A{DIM_X + DIM_V};       ///< @brief Augmented state dimension
+   static constexpr int32_t SIGMA_DIM_A{2 * DIM_A + 1}; ///< @brief Sigma points dimension for augmented state
+   float32_t m_kappa{0};                                ///< @brief Kappa parameter for finding sigma points
+
+   // Add variables to track the covariances of the process and measurement noise.
+   Matrix<DIM_V, DIM_V> m_matQ; // @brief Process noise covariance matrix
+   Matrix<DIM_N, DIM_N> m_matR; // @brief Measurement noise covariance matrix
+
+   Matrix<DIM_A, SIGMA_DIM_A> m_matSigmaXa{Matrix<DIM_A, SIGMA_DIM_A>::Zero()}; ///< @brief Sigma points matrix
+
+   TrackFitterUKF()
+      : KalmanFilter<DIM_X, DIM_Z>(), m_kappa(3 - DIM_A), m_matQ(Matrix<DIM_V, DIM_V>::Zero()),
+        m_matR(Matrix<DIM_N, DIM_N>::Zero())
+   {
+      // 1. calculate weights
+      updateWeights();
+   }
+
+   ~TrackFitterUKF() {}
+
+   void setKappa(float32_t kappa)
+   {
+      m_kappa = kappa; // Set the kappa parameter for sigma point calculation
+      updateWeights(); // Update the weights based on the new kappa value
+   }
+
+   // This code uses two different conventions for managing noise.
+   // The state vector noise is set in the updateAugmentedStateAndCovariance() method, while
+   // the noise vectors for the process and measurement models are set in the setCovarianceQ() and
+   // setCovarianceR() methods. This is an odd choice. We will be moving everything into a common
+   // structure where updateAugmentedStateAndCovariance() handle all covariance updates that are actually
+   // part of the augmented state vector.
+
+   ///
+   /// @brief adding process noise covariance Q to the augmented state covariance
+   /// matPa in the middle element of the diagonal.
+   ///
+   void setCovarianceQ(const Matrix<DIM_V, DIM_V> &matQ)
+   {
+      m_matQ = matQ; // Store the process noise covariance matrix
+   }
```

Comment:
Will be removed or replaced by a function call. This matrix changes each iteration (will be part of refactor, for now it's fine)